// @looker/visualizations-adapters/__mocks__/mockSdk.js
import "core-js/modules/es.object.to-string.js";
import "core-js/modules/es.promise.js";
import { mockSdkConfigResponse, mockSdkFieldsResponse, mockSdkDataResponse, mockSDKModelExploreResponse, mockSdkColorCollectionResponse } from './';
export var mockSDK = {
  color_collection: function color_collection() {
    return Promise.resolve({
      ok: true,
      value: mockSdkColorCollectionResponse
    });
  },
  create_query: function create_query() {
    return Promise.resolve({
      ok: true,
      value: {
        id: 126,
        vis_config: mockSdkConfigResponse,
        model: 'thelook',
        view: 'orders'
      }
    });
  },
  dashboard: function dashboard() {
    return Promise.resolve({
      ok: true,
      value: {
        dashboard_elements: [{
          query: {
            id: 126,
            vis_config: mockSdkConfigResponse,
            model: 'thelook',
            view: 'orders'
          }
        }]
      }
    });
  },
  lookml_model_explore: function lookml_model_explore() {
    return Promise.resolve({
      ok: true,
      value: mockSDKModelExploreResponse
    });
  },
  query_for_slug: function query_for_slug() {
    return Promise.resolve({
      ok: true,
      value: {
        id: 126,
        vis_config: mockSdkConfigResponse,
        model: 'thelook',
        view: 'orders'
      }
    });
  },
  run_query: function run_query() {
    return Promise.resolve({
      ok: true,
      error: null,
      value: {
        data: mockSdkDataResponse,
        fields: mockSdkFieldsResponse
      }
    });
  },
  query: function query() {
    return Promise.resolve({
      ok: true,
      value: {
        vis_config: mockSdkConfigResponse,
        model: 'thelook',
        view: 'orders'
      }
    });
  }
};
