// @looker/visualizations-adapters/__mocks__/mockFields.js
export var mockFields = {
  measures: [{
    is_numeric: true,
    label: 'Orders Count',
    label_short: 'Count',
    name: 'orders.count',
    sortable: true,
    type: 'count_distinct',
    view: 'orders',
    view_label: 'Orders',
    value_format: null
  }, {
    is_numeric: true,
    label: 'Orders Average Total Amount of Order USD',
    label_short: 'Average Total Amount of Order USD',
    name: 'orders.average_total_amount_of_order_usd',
    sortable: true,
    type: 'average',
    view: 'orders',
    view_label: 'Orders',
    value_format: '$#,##0.00'
  }],
  dimensions: [{
    is_filter: false,
    is_fiscal: false,
    is_numeric: false,
    is_timeframe: true,
    label: 'Orders Created Date',
    label_short: 'Created Date',
    name: 'orders.created_date',
    sortable: true,
    type: 'date_date',
    view: 'orders',
    view_label: 'Orders',
    value_format: null
  }, {
    is_filter: false,
    is_fiscal: false,
    is_numeric: false,
    is_timeframe: true,
    label: 'Users State',
    label_short: 'State',
    name: 'users.state',
    sortable: true,
    type: 'string',
    view: 'users',
    view_label: 'Users',
    value_format: null
  }]
};
