// @looker/visualizations-adapters/VisWrapper/VisWrapper.js
import "core-js/modules/es.object.keys.js";
import "core-js/modules/es.array.index-of.js";
import "core-js/modules/es.symbol.js";
import "core-js/modules/es.object.assign.js";
var _excluded = ["legend"];

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React, { useContext, forwardRef } from 'react';
import styled, { css, ThemeContext } from 'styled-components';
import { ComponentsProvider } from '@looker/components';
var VisWrapperInternal = forwardRef(function (_ref, ref) {
  var legend = _ref.legend,
      props = _objectWithoutProperties(_ref, _excluded);

  var theme = useContext(ThemeContext);

  if (!theme) {
    return React.createElement(ComponentsProvider, null, React.createElement(VisWrapperInternal, _extends({}, props, {
      ref: ref
    })));
  }

  return React.createElement("div", _extends({}, props, {
    ref: ref
  }));
});
VisWrapperInternal.displayName = "VisWrapperInternal";

var flexDirection = function flexDirection(_ref2) {
  var legend = _ref2.legend;
  var POSITION_MAP = {
    top: 'column-reverse',
    right: 'row',
    left: 'row-reverse',
    bottom: 'column'
  };
  var POSITION = legend ? legend.position : 'bottom';
  return css(["flex-direction:", ";justify-content:", ";"], POSITION_MAP[POSITION], POSITION === 'left' ? "flex-end" : "flex-start");
};

export var VisWrapper = styled(VisWrapperInternal).withConfig({
  displayName: "VisWrapper",
  componentId: "sc-sssqk1-0"
})(["display:flex;flex:1;", " & > div > svg{overflow:visible;}"], flexDirection);
