// @looker/visualizations-single-value/SingleValue/SingleValue.js
import "core-js/modules/es.function.name.js";
import "core-js/modules/es.array.is-array.js";
import React from 'react';
import numeral from 'numeral';
import styled from 'styled-components';
import { VisWrapper } from '@looker/visualizations-adapters';
export var SingleValue = function SingleValue(_ref) {
  var data = _ref.data,
      fields = _ref.fields,
      width = _ref.width,
      height = _ref.height,
      config = _ref.config;

  if (!data) {
    return null;
  }

  var _config$series = config.series,
      series = _config$series === void 0 ? {} : _config$series;
  var name = fields.measures[0].name;
  var firstSeries = Array.isArray(series) ? series[0] : series[name || ''];
  var color = firstSeries.color,
      label = firstSeries.label,
      value_format = firstSeries.value_format;
  var value = data[0][name || ''];
  var formattedValue = numeral(value).format(value_format);
  return React.createElement(VisWrapper, null, React.createElement(SingleValueLayout, {
    width: width,
    height: height
  }, React.createElement(SingleValueContent, {
    color: color
  }, formattedValue), label && React.createElement(SingleValueTitle, {
    color: color
  }, label)));
};
var SingleValueLayout = styled.div.withConfig({
  displayName: "SingleValue__SingleValueLayout",
  componentId: "sc-1qwcp7y-0"
})(["align-items:center;background:", ";display:grid;grid-template-columns:1fr;height:", ";justify-items:center;width:", ";"], function (_ref2) {
  var theme = _ref2.theme;
  return theme.colors.background;
}, function (_ref3) {
  var height = _ref3.height;
  return "".concat(height, "px") || "auto";
}, function (_ref4) {
  var width = _ref4.width;
  return "".concat(width, "px") || "auto";
});
var SingleValueContent = styled.div.withConfig({
  displayName: "SingleValue__SingleValueContent",
  componentId: "sc-1qwcp7y-1"
})(["color:", ";font-size:", ";"], function (_ref5) {
  var color = _ref5.color;
  return "".concat(color);
}, function (_ref6) {
  var theme = _ref6.theme;
  return theme.fontSizes.xxxlarge;
});
var SingleValueTitle = styled.div.withConfig({
  displayName: "SingleValue__SingleValueTitle",
  componentId: "sc-1qwcp7y-2"
})(["color:", ";font-size:", ";opacity:50%;"], function (_ref7) {
  var color = _ref7.color;
  return "".concat(color);
}, function (_ref8) {
  var theme = _ref8.theme;
  return theme.fontSizes.large;
});
