// @looker/components-date/FieldTime/FieldTime.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import "core-js/modules/es.array.from.js";
import "core-js/modules/es.string.iterator.js";
import "core-js/modules/es.array.some.js";
import "core-js/modules/es.object.assign.js";
import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { FloatingLabelField, getHasValue, omitFieldProps, pickFieldProps, useFormContext, useID } from '@looker/components';
import { InputTime } from '../InputTime';

var checkValueOnBlur = function checkValueOnBlur(e) {
  var target = e.currentTarget;
  var inputs = Array.from(target.querySelectorAll('input'));
  return inputs.some(function (input) {
    return input.value !== '';
  });
};

var FieldTimeComponent = forwardRef(function (props, ref) {
  var validationMessage = useFormContext(props);
  var id = useID(props.id);
  return React.createElement(FloatingLabelField, _extends({}, pickFieldProps(props), {
    id: id,
    validationMessage: validationMessage,
    hasValue: getHasValue(props),
    checkValueOnBlur: checkValueOnBlur
  }), React.createElement(InputTime, _extends({}, omitFieldProps(props), {
    width: "100%",
    "aria-describedby": "describedby-".concat(id),
    "aria-labelledby": "labelledby-".concat(id),
    id: id,
    validationType: validationMessage && validationMessage.type,
    ref: ref
  })));
});
FieldTimeComponent.displayName = 'FieldTimeComponent';
export var FieldTime = styled(FieldTimeComponent).withConfig({
  displayName: "FieldTime",
  componentId: "sc-towj3a-0"
})([""]);
