// @looker/components-date/FieldDateRange/FieldDateRange.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import "core-js/modules/es.array.from.js";
import "core-js/modules/es.string.iterator.js";
import "core-js/modules/es.array.some.js";
import "core-js/modules/es.object.assign.js";
import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { FloatingLabelField, omitFieldProps, pickFieldProps, useFormContext, useID } from '@looker/components';
import { InputDateRange } from '../InputDateRange';

var checkValueOnBlur = function checkValueOnBlur(e) {
  var inputs = Array.from(e.currentTarget.querySelectorAll('input'));
  return inputs.some(function (input) {
    return input.value !== '';
  });
};

var FieldDateRangeComponent = forwardRef(function (props, ref) {
  var validationMessage = useFormContext(props);
  var id = useID(props.id);
  return React.createElement(FloatingLabelField, _extends({}, pickFieldProps(props), {
    checkValueOnBlur: checkValueOnBlur,
    hasValue: !!props.defaultValue || !!props.value,
    id: id,
    validationMessage: validationMessage
  }), React.createElement(InputDateRange, _extends({}, omitFieldProps(props), {
    "aria-describedby": "describedby-".concat(id),
    "aria-labelledby": "labelledby-".concat(id),
    id: id,
    validationType: validationMessage && validationMessage.type,
    ref: ref
  })));
});
FieldDateRangeComponent.displayName = 'FieldDateRangeComponent';
export var FieldDateRange = styled(FieldDateRangeComponent).withConfig({
  displayName: "FieldDateRange",
  componentId: "sc-18dhwcx-0"
})([""]);
