// @looker/visualizations-visx/Scatter/Scatter.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import "core-js/modules/es.array.map.js";
import "core-js/modules/es.function.name.js";
import "core-js/modules/es.array.concat.js";
import "core-js/modules/es.object.define-property.js";
import "core-js/modules/es.object.keys.js";
import "core-js/modules/es.symbol.js";
import "core-js/modules/es.array.filter.js";
import "core-js/modules/es.object.get-own-property-descriptor.js";
import "core-js/modules/es.array.for-each.js";
import "core-js/modules/web.dom-collections.for-each.js";
import "core-js/modules/es.object.get-own-property-descriptors.js";
import "core-js/modules/es.object.define-properties.js";
import React, { Fragment } from 'react';
import { DataProvider, GlyphSeries, XYChart } from '@visx/xychart';
import { DEFAULT_HEIGHT, VisWrapper, hexToRgba, getSeriesMax, getSeriesMin } from '@looker/visualizations-adapters';
import { XYLegend } from '../XYLegend';
import isArray from 'lodash/isArray';
import get from 'lodash/get';
import { concatDimensions, getX, getY, getYAxisRange, useChartTheme, useAxis, isValidChartData, getRelativeGlyphSize, getDefaultGlyphSize, dimensionToDate, isDateQuery } from '../utils';
import { XYTooltip } from '../XYTooltip';
import { Glyph } from '../Glyph';
import { Grid } from '../Grid';
export var Scatter = function Scatter(_ref) {
  var data = _ref.data,
      config = _ref.config,
      _ref$height = _ref.height,
      height = _ref$height === void 0 ? DEFAULT_HEIGHT : _ref$height,
      width = _ref.width,
      fields = _ref.fields;
  var formattedData = concatDimensions(data, fields);

  var _useAxis = useAxis({
    config: config,
    data: formattedData,
    fields: fields
  }),
      XAxis = _useAxis.XAxis,
      YAxis = _useAxis.YAxis,
      chartMargin = _useAxis.chartMargin;

  var chartTheme = useChartTheme(config.series);

  if (!isValidChartData(data, fields)) {
    return null;
  }

  var plots = fields.measures.map(function (measure, i) {
    var series = isArray(config.series) ? get(config, ['series', i]) : get(config, ['series', measure.name]);
    if (!series.visible) return React.createElement(Fragment, {
      key: i
    });
    var style = series.style,
        shape = series.shape,
        size_by = series.size_by,
        _series$line_width = series.line_width,
        line_width = _series$line_width === void 0 ? 1 : _series$line_width;
    var id = "marker-".concat(shape, "-").concat(i);
    var sizeByMax = size_by ? getSeriesMax(size_by, data) : 0;
    var sizeByMin = size_by ? getSeriesMin(size_by, data) : 0;
    return React.createElement(Fragment, {
      key: id
    }, React.createElement(GlyphSeries, {
      key: id,
      dataKey: measure.name,
      data: dimensionToDate(formattedData, fields),
      xAccessor: function xAccessor(d) {
        return getX(d);
      },
      yAccessor: function yAccessor(d) {
        return getY(d, i);
      },
      size: function size(sizeProps) {
        if (size_by && sizeByMax - sizeByMin > 0) {
          return getRelativeGlyphSize(sizeProps[size_by], sizeByMin, sizeByMax);
        } else {
          return getDefaultGlyphSize(line_width);
        }
      },
      renderGlyph: function renderGlyph(_ref2) {
        var size = _ref2.size,
            color = _ref2.color,
            x = _ref2.x,
            y = _ref2.y;
        return React.createElement(Glyph, {
          series: series,
          top: y,
          left: x,
          size: size,
          fill: hexToRgba(color, 0.6),
          stroke: style === 'filled' ? false : undefined
        });
      }
    }));
  });
  var domain = getYAxisRange({
    config: config,
    data: formattedData,
    fields: fields
  });

  var Y_SCALE = _objectSpread({
    type: 'linear'
  }, domain && {
    domain: domain,
    zero: false
  });

  return React.createElement(DataProvider, {
    xScale: {
      type: isDateQuery(fields) ? 'time' : 'band'
    },
    yScale: Y_SCALE,
    theme: chartTheme
  }, React.createElement(VisWrapper, {
    legend: config.legend
  }, React.createElement(XYChart, {
    margin: chartMargin,
    width: width,
    height: height
  }, React.createElement(XAxis, null), React.createElement(YAxis, null), React.createElement(Grid, {
    config: config
  }), React.createElement(XYTooltip, {
    config: config,
    data: formattedData,
    fields: fields
  }), plots), React.createElement(XYLegend, {
    chartWidth: width,
    config: config,
    fields: fields
  })));
};
