// @looker/components-date/cjs/Calendar/CalendarNav.js
"use strict";

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/es.weak-map.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarNav = exports.headingSizeMap = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _noop = _interopRequireDefault(require("lodash/noop"));

var _ChevronLeft = require("@styled-icons/material-rounded/ChevronLeft");

var _ChevronRight = require("@styled-icons/material-rounded/ChevronRight");

var _components = require("@looker/components");

var _CalendarContext = require("./CalendarContext");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const headingSizeMap = size => {
  switch (size) {
    case 'small':
      return 'h6';

    case 'medium':
      return 'h5';

    case 'large':
      return 'h4';

    default:
      return 'h5';
  }
};

exports.headingSizeMap = headingSizeMap;

const CalendarNav = ({
  children,
  labels,
  nextMonth,
  previousMonth
}) => {
  const _useContext = (0, _react.useContext)(_CalendarContext.CalendarContext),
        _useContext$onNextCli = _useContext.onNextClick,
        onNextClick = _useContext$onNextCli === void 0 ? _noop.default : _useContext$onNextCli,
        _useContext$onPrevCli = _useContext.onPrevClick,
        onPrevClick = _useContext$onPrevCli === void 0 ? _noop.default : _useContext$onPrevCli,
        showPreviousButton = _useContext.showPreviousButton,
        showNextButton = _useContext.showNextButton,
        size = _useContext.size;

  const handleNextClick = e => {
    e.preventDefault();
    onNextClick(nextMonth);
  };

  const handlePreviousClick = e => {
    e.preventDefault();
    onPrevClick(previousMonth);
  };

  return _react.default.createElement(NavGrid, null, _react.default.createElement(NextButtonWrapper, null, showPreviousButton && _react.default.createElement(_components.IconButton, {
    icon: _react.default.createElement(_ChevronLeft.ChevronLeft, null),
    label: labels === null || labels === void 0 ? void 0 : labels.previousMonth,
    size: size,
    onClick: handlePreviousClick
  })), children, _react.default.createElement(PrevButtonWrapper, null, showNextButton && _react.default.createElement(_components.IconButton, {
    icon: _react.default.createElement(_ChevronRight.ChevronRight, null),
    label: labels === null || labels === void 0 ? void 0 : labels.nextMonth,
    size: size,
    onClick: handleNextClick,
    style: {
      justifySelf: 'right'
    }
  })));
};

exports.CalendarNav = CalendarNav;

const NavGrid = _styledComponents.default.div.withConfig({
  displayName: "CalendarNav__NavGrid",
  componentId: "sc-3k2ono-0"
})(["align-items:center;display:grid;grid-template-columns:40px 1fr 40px;justify-items:center;"]);

const NextButtonWrapper = _styledComponents.default.div.withConfig({
  displayName: "CalendarNav__NextButtonWrapper",
  componentId: "sc-3k2ono-1"
})(["justify-self:left;"]);

const PrevButtonWrapper = _styledComponents.default.div.withConfig({
  displayName: "CalendarNav__PrevButtonWrapper",
  componentId: "sc-3k2ono-2"
})(["justify-self:right;"]);
