// @looker/components-date/FieldDate/FieldDate.js
import "core-js/modules/es.object.assign.js";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { FloatingLabelField, omitFieldProps, pickFieldProps, useFormContext, useID } from '@looker/components';
import { InputDate } from '../InputDate';
var FieldDateComponent = forwardRef(function (props, ref) {
  var validationMessage = useFormContext(props);
  var id = useID(props.id);
  return React.createElement(FloatingLabelField, _extends({}, pickFieldProps(props), {
    hasValue: !!props.defaultValue || !!props.value,
    id: id,
    validationMessage: validationMessage
  }), React.createElement(InputDate, _extends({}, omitFieldProps(props), {
    "aria-describedby": "describedby-".concat(id),
    "aria-labelledby": "labelledby-".concat(id),
    id: id,
    validationType: validationMessage && validationMessage.type,
    ref: ref
  })));
});
FieldDateComponent.displayName = 'FieldDateComponent';
export var FieldDate = styled(FieldDateComponent).withConfig({
  displayName: "FieldDate",
  componentId: "sc-h19tmj-0"
})([""]);
