// @looker/components/Calendar/YearList.js
import "core-js/modules/es.object.keys.js";
import "core-js/modules/es.array.index-of.js";
import "core-js/modules/es.symbol.js";
var _excluded = ["baseMonth", "setBaseMonth", "onMonthChange", "currentDate"];

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import { getYear, setYear } from 'date-fns';
import React from 'react';
import { ScrollableDateList } from './ScrollableDateList';
import { Year } from './Year';

var getItemMonth = function getItemMonth(baseDate, offset) {
  var currentMonthIndex = getYear(baseDate);
  return setYear(baseDate, currentMonthIndex + offset);
};

export var YearList = function YearList(_ref) {
  var baseMonth = _ref.baseMonth,
      setBaseMonth = _ref.setBaseMonth,
      onMonthChange = _ref.onMonthChange,
      currentDate = _ref.currentDate,
      props = _objectWithoutProperties(_ref, _excluded);

  return React.createElement(ScrollableDateList, {
    currentDate: currentDate,
    baseMonth: baseMonth,
    setBaseMonth: setBaseMonth,
    onMonthChange: onMonthChange,
    buffer: 10,
    getItemDate: getItemMonth,
    itemComponent: Year,
    itemProps: props,
    thresholdRatio: 7 / 11
  });
};
