// @looker/visualizations-adapters/configTransformations/chartConfigByType.js
import "core-js/modules/es.array.concat.js";
import { barPositioning } from './barPositioning';
import { linePositioning } from './linePositioning';
import { legendPosition } from './legendPosition';
import { legendType } from './legendType';
import { legendValue } from './legendValue';
import { renderNullValues } from './renderNullValues';
import { seriesCellVis } from './seriesCellVis';
import { seriesColors } from './seriesColors';
import { seriesLabels } from './seriesLabels';
import { seriesLineWidth } from './seriesLineWidth';
import { dimensionSeriesColors } from './dimensionSeriesColors';
import { seriesPointShape } from './seriesPointShape';
import { seriesPointStyle } from './seriesPointStyle';
import { seriesSize } from './seriesSize';
import { seriesVisible } from './seriesVisible';
import { tooltips } from './tooltips';
import { truncateText } from './truncateText';
import { xAxis } from './xAxis';
import { yAxis } from './yAxis';
import { yAxisRange } from './yAxisRange';
import { seriesValueFormat } from './seriesValueFormat';
export var commonCartesianDefaults = [seriesLabels, seriesColors, seriesValueFormat, seriesVisible, legendPosition, tooltips, xAxis, yAxis];
export var commonLineDefaults = [seriesPointStyle, seriesPointShape, renderNullValues];
export var chartConfigByType = {
  area: [].concat(commonLineDefaults, commonCartesianDefaults, [linePositioning, seriesLineWidth]),
  bar: [barPositioning].concat(commonCartesianDefaults),
  column: [barPositioning].concat(commonCartesianDefaults),
  "default": [].concat(commonCartesianDefaults),
  table: [seriesCellVis, seriesLabels, seriesVisible, truncateText, seriesValueFormat, seriesColors],
  line: [].concat(commonLineDefaults, commonCartesianDefaults, [seriesLineWidth]),
  pie: [legendPosition, legendType, legendValue, dimensionSeriesColors, tooltips],
  scatter: [].concat(commonLineDefaults, commonCartesianDefaults, [seriesLineWidth, seriesSize, renderNullValues]),
  single_value: [seriesLabels, seriesColors, seriesValueFormat],
  sparkline: [seriesColors, seriesLineWidth, renderNullValues, yAxisRange]
};
