// @looker/visualizations-visx/Pie/PieTooltip.js
import "core-js/modules/es.function.name.js";
import React from 'react';
import { TooltipWithBounds } from '@visx/tooltip';
import { SpaceVertical } from '@looker/components';
import styled from 'styled-components';
import { tooltipStyles } from '../XYTooltip';
import { DLGroup } from '../DLGroup';
export var PieTooltip = styled(function (_ref) {
  var tooltipOpen = _ref.tooltipOpen,
      tooltipData = _ref.tooltipData,
      tooltipTop = _ref.tooltipTop,
      tooltipLeft = _ref.tooltipLeft,
      measure = _ref.measure,
      dimension = _ref.dimension,
      className = _ref.className;

  if (tooltipOpen && tooltipData) {
    return React.createElement(TooltipWithBounds, {
      className: className,
      top: tooltipTop,
      left: tooltipLeft,
      unstyled: true,
      applyPositionStyle: true
    }, React.createElement("dl", null, React.createElement(SpaceVertical, {
      gap: "u3"
    }, React.createElement(DLGroup, {
      value: tooltipData[dimension.name]
    }), React.createElement(DLGroup, {
      label: measure.label || measure.label_short,
      value: tooltipData[measure.name]
    }))));
  }

  return null;
}).withConfig({
  displayName: "PieTooltip",
  componentId: "sc-1vkcacc-0"
})(["", ""], tooltipStyles);
