// @looker/filter-expressions/cjs/utils/date/date_conversions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.hasTimeFilterDateTimeModel = exports.clearTimeFilterDateTimeModel = exports.addDays = exports.filterDateTimeModelToDate = exports.dateToFilterDateTimeModel = void 0;

const dateToFilterDateTimeModel = (date = new Date()) => ({
  year: date.getFullYear(),
  month: date.getMonth() + 1,
  day: date.getDate(),
  hour: date.getHours(),
  minute: date.getMinutes(),
  second: date.getSeconds()
});

exports.dateToFilterDateTimeModel = dateToFilterDateTimeModel;

const filterDateTimeModelToDate = ({
  year,
  month,
  day,
  hour = 0,
  minute = 0,
  second = 0
}) => new Date(year, month - 1, day, hour, minute, second);

exports.filterDateTimeModelToDate = filterDateTimeModelToDate;

const addDays = (date, days) => {
  const result = new Date(date);
  result.setDate(result.getDate() + days);
  return result;
};

exports.addDays = addDays;

const clearTimeFilterDateTimeModel = model => ({
  year: model.year,
  month: model.month,
  day: model.day
});

exports.clearTimeFilterDateTimeModel = clearTimeFilterDateTimeModel;

const hasTimeFilterDateTimeModel = model => model !== undefined && model.hour !== undefined && model.minute !== undefined && model.second !== undefined;

exports.hasTimeFilterDateTimeModel = hasTimeFilterDateTimeModel;
