// @looker/filter-expressions/utils/get_expression_type.js
import "core-js/modules/es.array.some.js";
import "core-js/modules/es.array.index-of.js";
import { Category } from '@looker/sdk';

var getTimeframeExpressionType = function getTimeframeExpressionType(fieldType) {
  if (!fieldType) return 'date';
  var isDateTime = ['date_time', 'hour', 'minute', 'second'].some(function (timeString) {
    return fieldType.indexOf(timeString) > -1;
  });
  if (isDateTime) return 'date_time';
  return 'date';
};

export var getExpressionTypeFromField = function getExpressionTypeFromField(field) {
  if ((field === null || field === void 0 ? void 0 : field.category) === Category.parameter && (field === null || field === void 0 ? void 0 : field.type) === 'number') {
    return field.type;
  }

  if (field.enumerations) {
    return 'tier';
  }

  if (field.is_numeric) {
    return 'number';
  }

  if (field.is_timeframe) {
    return getTimeframeExpressionType(field.type);
  }

  if (field.type === 'location' || field.type === 'location_bin_level') {
    return 'location';
  }

  return 'string';
};
export var getExpressionType = function getExpressionType(filter) {
  if (filter.field) {
    return getExpressionTypeFromField(filter.field);
  }

  var filterExpressionMap = {
    number_filter: 'number',
    string_filter: 'string',
    date_filter: 'date',
    field_filter: 'string'
  };
  return filterExpressionMap[filter.type];
};
