// @looker/components/cjs/Calendar/DaysOfWeek.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DaysOfWeek = void 0;

require("core-js/modules/es.array.from.js");

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/es.array.slice.js");

require("core-js/modules/es.array.map.js");

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Layout = require("../Layout");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const DaysOfWeek = (0, _styledComponents.default)(({
  className,
  firstDayOfWeek,
  locale
}) => {
  const allDays = Array.from(Array(7), (_, i) => {
    var _locale$localize;

    return (_locale$localize = locale.localize) === null || _locale$localize === void 0 ? void 0 : _locale$localize.day(i, {
      width: 'narrow'
    });
  });
  const daysOfWeek = [...allDays.slice(firstDayOfWeek), ...allDays.slice(0, firstDayOfWeek)];
  return _react.default.createElement("div", {
    className: className
  }, _react.default.createElement(_Layout.Grid, {
    columns: 7,
    gap: "none"
  }, daysOfWeek.map((day, i) => _react.default.createElement("div", {
    key: i
  }, day))));
}).withConfig({
  displayName: "DaysOfWeek",
  componentId: "sc-10k3m05-0"
})(["background:", ";", "{padding-left:", ";width:fit-content;div{color:", ";font-family:", ";font-size:", ";line-height:", ";text-align:center;width:", ";}}"], ({
  theme
}) => theme.colors.ui1, _Layout.Grid, ({
  theme
}) => theme.space.u5, ({
  theme
}) => theme.colors.text2, ({
  theme
}) => theme.fonts.body, ({
  theme
}) => theme.fontSizes.xsmall, ({
  theme
}) => theme.space.u8, ({
  theme
}) => theme.space.u8);
exports.DaysOfWeek = DaysOfWeek;
