// @looker/components/cjs/Calendar/MonthTitle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MonthTitle = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Text = require("../Text");

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const MonthTitle = (0, _styledComponents.default)(({
  className,
  locale,
  month
}) => {
  return _react.default.createElement(_Text.Span, {
    className: className,
    fontSize: "xsmall",
    color: "text2",
    fontWeight: "bold",
    px: "u8",
    py: "u2"
  }, (0, _utils.formatDateString)(month, 'MMM yyyy', locale));
}).withConfig({
  displayName: "MonthTitle",
  componentId: "sc-1cr2h3a-0"
})(["", " display:block;margin:", ";position:relative;z-index:", ";"], _utils.rangeTypeStyle, ({
  inline,
  theme: {
    space
  }
}) => inline ? `0 0 -${space.u8} 0` : `${space.u05} 0`, ({
  theme
}) => theme.zIndexFloor);
exports.MonthTitle = MonthTitle;
