// @looker/visualizations-visx/Column/Column.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import "core-js/modules/es.array.map.js";
import "core-js/modules/es.function.name.js";
import "core-js/modules/es.object.define-property.js";
import "core-js/modules/es.object.keys.js";
import "core-js/modules/es.symbol.js";
import "core-js/modules/es.array.filter.js";
import "core-js/modules/es.object.get-own-property-descriptor.js";
import "core-js/modules/es.array.for-each.js";
import "core-js/modules/web.dom-collections.for-each.js";
import "core-js/modules/es.object.get-own-property-descriptors.js";
import "core-js/modules/es.object.define-properties.js";
import React from 'react';
import { DataProvider, BarSeries, XYChart, BarStack, BarGroup } from '@visx/xychart';
import { DEFAULT_HEIGHT, VisWrapper } from '@looker/visualizations-adapters';
import { XYLegend } from '../XYLegend';
import isArray from 'lodash/isArray';
import get from 'lodash/get';
import compact from 'lodash/compact';
import { concatDimensions, getX, getY, getYAxisRange, useAxis, useChartTheme, isValidChartData } from '../utils';
import { XYTooltip } from '../XYTooltip';
import { Grid } from '../Grid';
export var Column = function Column(_ref) {
  var data = _ref.data,
      config = _ref.config,
      _ref$height = _ref.height,
      height = _ref$height === void 0 ? DEFAULT_HEIGHT : _ref$height,
      width = _ref.width,
      fields = _ref.fields;
  var positioning = config.positioning,
      seriesList = config.series,
      legend = config.legend;
  var formattedData = concatDimensions(data, fields);

  var _useAxis = useAxis({
    config: config,
    data: formattedData,
    fields: fields
  }),
      XAxis = _useAxis.XAxis,
      YAxis = _useAxis.YAxis,
      chartMargin = _useAxis.chartMargin;

  var chartTheme = useChartTheme(seriesList);

  if (!isValidChartData(data, fields)) {
    return null;
  }

  var renderedColumns = compact(fields.measures.map(function (measure, i) {
    var series = isArray(seriesList) ? get(config, ['series', i]) : get(config, ['series', measure.name]);
    if (!series.visible) return undefined;
    return React.createElement(BarSeries, {
      key: i,
      dataKey: measure.name,
      data: formattedData,
      xAccessor: function xAccessor(d) {
        return getX(d);
      },
      yAccessor: function yAccessor(d) {
        return getY(d, i);
      }
    });
  }));
  var domain = positioning === 'percent' ? [0, 1] : getYAxisRange({
    config: config,
    data: formattedData,
    fields: fields
  });

  var Y_SCALE = _objectSpread({
    type: 'linear'
  }, domain && {
    domain: domain,
    zero: false
  });

  return React.createElement(DataProvider, {
    xScale: {
      type: 'band',
      paddingInner: 0.2
    },
    yScale: Y_SCALE,
    theme: chartTheme
  }, React.createElement(VisWrapper, {
    legend: legend
  }, React.createElement(XYChart, {
    margin: chartMargin,
    width: width,
    height: height
  }, React.createElement(XAxis, null), React.createElement(YAxis, null), React.createElement(Grid, {
    config: config
  }), React.createElement(XYTooltip, {
    config: config,
    data: formattedData,
    fields: fields,
    showDatumGlyph: false,
    snapToDatum: false
  }), positioning === 'stacked' || positioning === 'percent' ? React.createElement(BarStack, {
    offset: positioning === 'percent' ? 'expand' : 'none'
  }, renderedColumns) : React.createElement(BarGroup, null, renderedColumns)), React.createElement(XYLegend, {
    chartWidth: width,
    config: config,
    fields: fields
  })));
};
