// @looker/filter-expressions/utils/string/sanitize_string.js
import "core-js/modules/es.object.define-property.js";
import "core-js/modules/es.object.keys.js";
import "core-js/modules/es.symbol.js";
import "core-js/modules/es.array.filter.js";
import "core-js/modules/es.object.get-own-property-descriptor.js";
import "core-js/modules/es.array.for-each.js";
import "core-js/modules/web.dom-collections.for-each.js";
import "core-js/modules/es.object.get-own-property-descriptors.js";
import "core-js/modules/es.object.define-properties.js";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { findUserAttribute } from '../user_attribute';
export var sanitizeString = function sanitizeString(item) {
  var userAttributes = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  var _item$id = item.id,
      id = _item$id === void 0 ? '0' : _item$id,
      _item$is = item.is,
      is = _item$is === void 0 ? true : _item$is,
      type = item.type,
      _item$value = item.value,
      value = _item$value === void 0 ? [] : _item$value,
      attributeName = item.attributeName;
  var userAttribute = findUserAttribute(attributeName, userAttributes);

  switch (type) {
    case 'match':
      return {
        id: id,
        is: is,
        type: type,
        value: userAttribute ? [userAttribute.value] : value,
        attributeName: '',
        attributeValue: ''
      };

    default:
      return _objectSpread(_objectSpread({}, item), {}, {
        attributeName: '',
        attributeValue: ''
      });
  }
};
