// @looker/visualizations-visx/cjs/utils/useChartTheme.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useChartTheme = void 0;

require("core-js/modules/es.object.values.js");

require("core-js/modules/es.array.map.js");

var _react = require("react");

var _styledComponents = require("styled-components");

var _isArray = _interopRequireDefault(require("lodash/isArray"));

var _compact = _interopRequireDefault(require("lodash/compact"));

var _xychart = require("@visx/xychart");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const useChartTheme = series => {
  const theme = (0, _react.useContext)(_styledComponents.ThemeContext);
  const seriesList = series && (0, _isArray.default)(series) ? series : Object.values(series || {});
  const scheme = (0, _compact.default)(seriesList.map(s => s.visible && s.color));
  const chartTheme = (0, _xychart.buildChartTheme)({
    backgroundColor: theme.colors.background,
    colors: scheme.length ? scheme : [theme.colors.key],
    gridColor: theme.colors.ui2,
    gridColorDark: theme.colors.text5,
    svgLabelBig: {
      fill: theme.colors.text5
    },
    svgLabelSmall: {
      fill: theme.colors.text5
    },
    tickLength: 8
  });
  return chartTheme;
};

exports.useChartTheme = useChartTheme;
