// @looker/components/Calendar/Cell.js
import React from 'react';
import styled, { css } from 'styled-components';
import { rangeTypeStyle } from './utils';

var rangePositionWidth = function rangePositionWidth(_ref) {
  var rangePosition = _ref.rangePosition,
      weekStart = _ref.weekStart,
      weekEnd = _ref.weekEnd;

  if (!rangePosition || rangePosition === 'middle') {
    return css(["width:100%;"]);
  }

  if (rangePosition === 'end' && weekStart || rangePosition === 'start' && weekEnd) {
    return css(["width:calc(100% - ", ");"], function (_ref2) {
      var space = _ref2.theme.space;
      return space.u4;
    });
  }

  return css(["width:", ";"], function (_ref3) {
    var theme = _ref3.theme;
    return theme.space.u4;
  });
};

var rangePositionLeft = function rangePositionLeft(_ref4) {
  var rangePosition = _ref4.rangePosition,
      weekStart = _ref4.weekStart;

  if (rangePosition === 'start') {
    if (weekStart) {
      return css(["left:auto;right:0;"]);
    } else {
      return css(["left:", ";"], function (_ref5) {
        var theme = _ref5.theme;
        return theme.space.u4;
      });
    }
  }

  return css(["left:0;"]);
};

export var Cell = styled(function (_ref6) {
  var children = _ref6.children,
      className = _ref6.className;
  return React.createElement("div", {
    className: className
  }, children);
}).withConfig({
  displayName: "Cell",
  componentId: "sc-y9lybp-0"
})(["margin:", " 0;padding-left:", ";padding-right:", ";position:relative;&::before{content:' ';height:100%;position:absolute;", " ", " ", "}"], function (_ref7) {
  var theme = _ref7.theme;
  return theme.space.u05;
}, function (_ref8) {
  var theme = _ref8.theme,
      weekStart = _ref8.weekStart;
  return weekStart ? theme.space.u5 : '0';
}, function (_ref9) {
  var theme = _ref9.theme,
      weekEnd = _ref9.weekEnd;
  return weekEnd ? theme.space.u5 : '0';
}, rangePositionWidth, rangePositionLeft, rangeTypeStyle);
