// @looker/components/Calendar/DaysOfWeek.js
function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

import "core-js/modules/es.array.from.js";
import "core-js/modules/es.string.iterator.js";
import "core-js/modules/es.array.concat.js";
import "core-js/modules/es.array.slice.js";
import "core-js/modules/es.array.map.js";
import "core-js/modules/es.array.is-array.js";
import "core-js/modules/es.symbol.js";
import "core-js/modules/es.symbol.description.js";
import "core-js/modules/es.object.to-string.js";
import "core-js/modules/es.symbol.iterator.js";
import "core-js/modules/es.array.iterator.js";
import "core-js/modules/web.dom-collections.iterator.js";
import "core-js/modules/es.function.name.js";
import React from 'react';
import styled from 'styled-components';
import { Grid } from '../Layout';
export var DaysOfWeek = styled(function (_ref) {
  var className = _ref.className,
      firstDayOfWeek = _ref.firstDayOfWeek,
      locale = _ref.locale;
  var allDays = Array.from(Array(7), function (_, i) {
    var _locale$localize;

    return (_locale$localize = locale.localize) === null || _locale$localize === void 0 ? void 0 : _locale$localize.day(i, {
      width: 'narrow'
    });
  });
  var daysOfWeek = [].concat(_toConsumableArray(allDays.slice(firstDayOfWeek)), _toConsumableArray(allDays.slice(0, firstDayOfWeek)));
  return React.createElement("div", {
    className: className
  }, React.createElement(Grid, {
    columns: 7,
    gap: "none"
  }, daysOfWeek.map(function (day, i) {
    return React.createElement("div", {
      key: i
    }, day);
  })));
}).withConfig({
  displayName: "DaysOfWeek",
  componentId: "sc-10k3m05-0"
})(["background:", ";", "{padding-left:", ";width:fit-content;div{color:", ";font-family:", ";font-size:", ";line-height:", ";text-align:center;width:", ";}}"], function (_ref2) {
  var theme = _ref2.theme;
  return theme.colors.ui1;
}, Grid, function (_ref3) {
  var theme = _ref3.theme;
  return theme.space.u5;
}, function (_ref4) {
  var theme = _ref4.theme;
  return theme.colors.text2;
}, function (_ref5) {
  var theme = _ref5.theme;
  return theme.fonts.body;
}, function (_ref6) {
  var theme = _ref6.theme;
  return theme.fontSizes.xsmall;
}, function (_ref7) {
  var theme = _ref7.theme;
  return theme.space.u8;
}, function (_ref8) {
  var theme = _ref8.theme;
  return theme.space.u8;
});
