// @looker/components/cjs/Calendar/Day.js
"use strict";

require("core-js/modules/es.array.index-of.js");

require("core-js/modules/es.array.filter.js");

require("core-js/modules/es.object.get-own-property-descriptors.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Day = exports.HitArea = void 0;

var _dateFns = require("date-fns");

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _pick = _interopRequireDefault(require("lodash/pick"));

var _Ripple = require("../Ripple");

var _utils = require("./utils");

const _excluded = ["className", "date", "locale", "onDraftSelect", "onSelect", "selected", "style"],
      _excluded2 = ["callbacks"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

const HitArea = _styledComponents.default.div.withConfig({
  displayName: "Day__HitArea",
  componentId: "sc-14dncin-0"
})(["height:", ";left:50%;position:absolute;top:50%;transform:translate(-50%,-50%);width:", ";"], ({
  theme
}) => theme.space.u8, ({
  theme
}) => theme.space.u8);

exports.HitArea = HitArea;
const Day = (0, _styledComponents.default)(_ref => {
  let className = _ref.className,
      date = _ref.date,
      locale = _ref.locale,
      onDraftSelect = _ref.onDraftSelect,
      onSelect = _ref.onSelect,
      selected = _ref.selected,
      style = _ref.style,
      props = _objectWithoutProperties(_ref, _excluded);

  const dateString = (0, _utils.formatDateString)(date, 'EEE MMM dd, yyyy', locale);
  const today = (0, _dateFns.isSameDay)(date, new Date());

  const handleClick = () => {
    onSelect(date);
  };

  const handleHoverFocus = () => onDraftSelect(date);

  const _useRipple = (0, _Ripple.useRipple)({
    className,
    color: 'neutral',
    style
  }),
        callbacks = _useRipple.callbacks,
        rippleProps = _objectWithoutProperties(_useRipple, _excluded2);

  const rippleHandlers = (0, _Ripple.useRippleHandlers)(callbacks, _objectSpread(_objectSpread({}, (0, _pick.default)(_objectSpread({}, props), _Ripple.rippleHandlerKeys)), {}, {
    onFocus: handleHoverFocus,
    onMouseEnter: handleHoverFocus
  }), props.disabled);
  return _react.default.createElement("button", _extends({}, props, {
    "aria-current": today ? 'date' : 'false',
    "aria-label": dateString,
    "aria-selected": selected,
    onClick: handleClick,
    title: dateString,
    type: "button"
  }, rippleProps, rippleHandlers), _react.default.createElement(HitArea, null), (0, _dateFns.getDate)(date));
}).withConfig({
  displayName: "Day",
  componentId: "sc-14dncin-1"
})(["", " background-color:transparent;border:none;border-radius:", ";box-sizing:border-box;color:", ";font-family:", ";font-size:", ";height:", ";line-height:", ";margin:0 ", ";outline:none;position:relative;width:", ";&[aria-current='date']{border:1px solid ", ";color:", ";}&[aria-selected='true']{background:", ";border:1px solid ", ";color:", ";}"], _Ripple.rippleStyle, ({
  theme
}) => theme.space.u5, ({
  theme
}) => theme.colors.text3, ({
  theme
}) => theme.fonts.body, ({
  theme
}) => theme.fontSizes.xsmall, ({
  theme
}) => theme.space.u7, ({
  theme
}) => theme.space.u4, ({
  theme
}) => theme.space.u05, ({
  theme
}) => theme.space.u7, ({
  theme
}) => theme.colors.key, ({
  theme
}) => theme.colors.key, ({
  theme
}) => theme.colors.key, ({
  theme
}) => theme.colors.key, ({
  theme
}) => theme.colors.background);
exports.Day = Day;
