// @looker/visualizations-adapters/cjs/__mocks__/mockConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mockScatterConfig = exports.mockTableConfig = exports.mockLineConfig = exports.mockPieConfig = exports.mockBarConfig = void 0;
const mockBarConfig = {
  type: 'bar',
  positioning: 'grouped',
  legend: {
    position: 'bottom'
  },
  series: [{
    color: '#ffcccc',
    visible: true
  }, {
    color: '#00ffcc',
    visible: true
  }],
  x_axis: [{
    reversed: false,
    label: 'Orders',
    values: true,
    gridlines: true
  }],
  y_axis: [{
    label: false,
    values: true,
    gridlines: true,
    range: ['auto', 'auto']
  }]
};
exports.mockBarConfig = mockBarConfig;
const mockPieConfig = {
  type: 'pie',
  legend: {
    position: 'bottom',
    value: 'label',
    type: 'legend'
  },
  series: [{
    color: '#fa8072'
  }]
};
exports.mockPieConfig = mockPieConfig;
const mockLineConfig = {
  type: 'line',
  legend: {
    position: 'bottom'
  },
  series: [{
    color: '#ff0000',
    style: 'filled'
  }, {
    color: '#00FF00',
    shape: 'diamond'
  }],
  y_axis: [{
    range: ['auto', 'auto']
  }]
};
exports.mockLineConfig = mockLineConfig;
const mockTableConfig = {
  type: 'table'
};
exports.mockTableConfig = mockTableConfig;
const mockScatterConfig = {
  type: 'scatter'
};
exports.mockScatterConfig = mockScatterConfig;
