// @looker/visualizations-adapters/cjs/__mocks__/mockSdk.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mockSDK = void 0;

require("core-js/modules/es.promise.js");

var _ = require("./");

const mockSDK = {
  color_collection: () => Promise.resolve({
    ok: true,
    value: _.mockSdkColorCollectionResponse
  }),
  create_query: () => Promise.resolve({
    ok: true,
    value: {
      id: 126,
      vis_config: _.mockSdkConfigResponse,
      model: 'thelook',
      view: 'orders'
    }
  }),
  dashboard: () => Promise.resolve({
    ok: true,
    value: {
      dashboard_elements: [{
        query: {
          id: 126,
          vis_config: _.mockSdkConfigResponse,
          model: 'thelook',
          view: 'orders'
        }
      }]
    }
  }),
  lookml_model_explore: () => Promise.resolve({
    ok: true,
    value: _.mockSDKModelExploreResponse
  }),
  query_for_slug: () => Promise.resolve({
    ok: true,
    value: {
      id: 126,
      vis_config: _.mockSdkConfigResponse,
      model: 'thelook',
      view: 'orders'
    }
  }),
  run_query: () => Promise.resolve({
    ok: true,
    error: null,
    value: {
      data: _.mockSdkDataResponse,
      fields: _.mockSdkFieldsResponse
    }
  }),
  query: () => Promise.resolve({
    ok: true,
    value: {
      vis_config: _.mockSdkConfigResponse,
      model: 'thelook',
      view: 'orders'
    }
  })
};
exports.mockSDK = mockSDK;
