// @looker/visualizations-visx/cjs/Pie/PieTooltip.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PieTooltip = void 0;

var _react = _interopRequireDefault(require("react"));

var _tooltip = require("@visx/tooltip");

var _components = require("@looker/components");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _XYTooltip = require("../XYTooltip");

var _DLGroup = require("../DLGroup");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const PieTooltip = (0, _styledComponents.default)(({
  tooltipOpen,
  tooltipData,
  tooltipTop,
  tooltipLeft,
  measure,
  dimension,
  className
}) => {
  if (tooltipOpen && tooltipData) {
    return _react.default.createElement(_tooltip.TooltipWithBounds, {
      className: className,
      top: tooltipTop,
      left: tooltipLeft,
      unstyled: true,
      applyPositionStyle: true
    }, _react.default.createElement("dl", null, _react.default.createElement(_components.SpaceVertical, {
      gap: "u3"
    }, _react.default.createElement(_DLGroup.DLGroup, {
      value: tooltipData[dimension.name]
    }), _react.default.createElement(_DLGroup.DLGroup, {
      label: measure.label || measure.label_short,
      value: tooltipData[measure.name]
    }))));
  }

  return null;
}).withConfig({
  displayName: "PieTooltip",
  componentId: "sc-1vkcacc-0"
})(["", ""], _XYTooltip.tooltipStyles);
exports.PieTooltip = PieTooltip;
