// @visx/drag/lib/useDrag.js
"use strict";

exports.__esModule = true;
exports.default = useDrag;

var _react = require("react");

var _event = require("@visx/event");

var _useStateWithCallback2 = _interopRequireDefault(require("./util/useStateWithCallback"));

var _clampNumber = _interopRequireDefault(require("./util/clampNumber"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/** Hook for dragging, returns a `UseDrag` object. */
function useDrag(_temp) {
  var _ref = _temp === void 0 ? {} : _temp,
      _ref$resetOnStart = _ref.resetOnStart,
      resetOnStart = _ref$resetOnStart === void 0 ? false : _ref$resetOnStart,
      _ref$snapToPointer = _ref.snapToPointer,
      snapToPointer = _ref$snapToPointer === void 0 ? true : _ref$snapToPointer,
      onDragEnd = _ref.onDragEnd,
      onDragMove = _ref.onDragMove,
      onDragStart = _ref.onDragStart,
      x = _ref.x,
      y = _ref.y,
      dx = _ref.dx,
      dy = _ref.dy,
      isDragging = _ref.isDragging,
      _ref$restrict = _ref.restrict,
      restrict = _ref$restrict === void 0 ? {} : _ref$restrict;

  // use ref to detect prop changes
  var positionPropsRef = (0, _react.useRef)({
    x: x,
    y: y,
    dx: dx,
    dy: dy
  });
  var xMin = restrict.xMin,
      xMax = restrict.xMax,
      yMin = restrict.yMin,
      yMax = restrict.yMax;
  var clampX = (0, _react.useCallback)(function (num) {
    return (0, _clampNumber.default)(num, xMin != null ? xMin : -Infinity, xMax != null ? xMax : Infinity);
  }, [xMax, xMin]);
  var clampY = (0, _react.useCallback)(function (num) {
    return (0, _clampNumber.default)(num, yMin != null ? yMin : -Infinity, yMax != null ? yMax : Infinity);
  }, [yMax, yMin]);

  var _useStateWithCallback = (0, _useStateWithCallback2.default)({
    x: x,
    y: y,
    dx: dx != null ? dx : 0,
    dy: dy != null ? dy : 0,
    isDragging: false
  }),
      dragState = _useStateWithCallback[0],
      setDragStateWithCallback = _useStateWithCallback[1]; // Track distance between pointer on dragStart and position of element being dragged


  var _useState = (0, _react.useState)({
    x: 0,
    y: 0
  }),
      dragStartPointerOffset = _useState[0],
      setDragStartPointerOffset = _useState[1]; // if prop position changes, update state


  (0, _react.useEffect)(function () {
    if (positionPropsRef.current.x !== x || positionPropsRef.current.y !== y || positionPropsRef.current.dx !== dx || positionPropsRef.current.dy !== dy) {
      positionPropsRef.current = {
        x: x,
        y: y,
        dx: dx,
        dy: dy
      };
      setDragStateWithCallback(function (currState) {
        return _extends({}, currState, {
          x: x,
          y: y,
          dx: dx != null ? dx : 0,
          dy: dy != null ? dy : 0
        });
      });
    }
  });
  (0, _react.useEffect)(function () {
    if (isDragging !== undefined && dragState.isDragging !== isDragging) {
      setDragStateWithCallback(function (currState) {
        return _extends({}, currState, {
          isDragging: isDragging
        });
      });
    }
  }, [dragState.isDragging, isDragging, setDragStateWithCallback]);
  var handleDragStart = (0, _react.useCallback)(function (event) {
    event.persist();
    setDragStateWithCallback(function (currState) {
      var currentPoint = {
        x: (currState.x || 0) + currState.dx,
        y: (currState.y || 0) + currState.dy
      };
      var eventPoint = (0, _event.localPoint)(event) || {
        x: 0,
        y: 0
      };
      var point = snapToPointer ? eventPoint : currentPoint;
      setDragStartPointerOffset({
        x: currentPoint.x - eventPoint.x,
        y: currentPoint.y - eventPoint.y
      });
      return {
        isDragging: true,
        dx: resetOnStart ? 0 : currState.dx,
        dy: resetOnStart ? 0 : currState.dy,
        x: resetOnStart ? clampX(point.x) : clampX(point.x) - currState.dx,
        y: resetOnStart ? clampY(point.y) : clampY(point.y) - currState.dy
      };
    }, onDragStart && function (currState) {
      onDragStart(_extends({}, currState, {
        event: event
      }));
    });
  }, [clampX, clampY, onDragStart, resetOnStart, setDragStateWithCallback, snapToPointer]);
  var handleDragMove = (0, _react.useCallback)(function (event) {
    event.persist();
    setDragStateWithCallback(function (currState) {
      var point = (0, _event.localPoint)(event) || {
        x: 0,
        y: 0
      };
      return currState.isDragging ? _extends({}, currState, {
        isDragging: true,
        dx: snapToPointer ? clampX(point.x) - (currState.x || 0) : clampX(point.x + dragStartPointerOffset.x) - (currState.x || 0),
        dy: snapToPointer ? clampY(point.y) - (currState.y || 0) : clampY(point.y + dragStartPointerOffset.y) - (currState.y || 0)
      }) : currState;
    }, onDragMove && function (currState) {
      if (currState.isDragging) onDragMove(_extends({}, currState, {
        event: event
      }));
    });
  }, [setDragStateWithCallback, onDragMove, snapToPointer, dragStartPointerOffset.x, dragStartPointerOffset.y, clampX, clampY]);
  var handleDragEnd = (0, _react.useCallback)(function (event) {
    event.persist();
    setDragStateWithCallback(function (currState) {
      return _extends({}, currState, {
        isDragging: false
      });
    }, onDragEnd && function (currState) {
      onDragEnd(_extends({}, currState, {
        event: event
      }));
    });
  }, [onDragEnd, setDragStateWithCallback]);
  return _extends({}, dragState, {
    dragEnd: handleDragEnd,
    dragMove: handleDragMove,
    dragStart: handleDragStart
  });
}
