// @looker/components-data/hooks/useQueryMetadata.js
import "core-js/modules/es.object.to-string.js";
import "core-js/modules/es.promise.js";
import "core-js/modules/es.object.define-property.js";
import "core-js/modules/es.object.keys.js";
import "core-js/modules/es.symbol.js";
import "core-js/modules/es.array.filter.js";
import "core-js/modules/es.object.get-own-property-descriptor.js";
import "core-js/modules/es.array.for-each.js";
import "core-js/modules/web.dom-collections.for-each.js";
import "core-js/modules/es.object.get-own-property-descriptors.js";
import "core-js/modules/es.object.define-properties.js";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import "regenerator-runtime/runtime.js";

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

import { useEffect, useMemo } from 'react';
import isEmpty from 'lodash/isEmpty';
import isEqual from 'lodash/isEqual';
import memoize from 'lodash/memoize';
import useSWR from 'swr';
import { getErrorResponse, isSuccessResponse } from '../utils';
import { useSDK } from './useSDK';
import { DataState } from './useDataState';
var fetchQueryMetadata = memoize(function () {
  var _ref = _asyncToGenerator(regeneratorRuntime.mark(function _callee(requestId, sdk) {
    var result;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return sdk.query(String(requestId));

          case 2:
            result = _context.sent;
            return _context.abrupt("return", result);

          case 4:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function (_x, _x2) {
    return _ref.apply(this, arguments);
  };
}());
export var useQueryMetadata = function useQueryMetadata(id) {
  var sdk = useSDK();

  var _DataState$useContain = DataState.useContainer(),
      getById = _DataState$useContain.getById,
      setById = _DataState$useContain.setById;

  var metadata = useMemo(function () {
    return getById(id, 'metadata') || {};
  }, [id, getById]);

  var fetcher = function () {
    var _ref2 = _asyncToGenerator(regeneratorRuntime.mark(function _callee2() {
      return regeneratorRuntime.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              if (!(id > 0 && (isEmpty(metadata.vis_config) || !metadata.model || !metadata.view))) {
                _context2.next = 2;
                break;
              }

              return _context2.abrupt("return", fetchQueryMetadata(id, sdk));

            case 2:
              return _context2.abrupt("return", undefined);

            case 3:
            case "end":
              return _context2.stop();
          }
        }
      }, _callee2);
    }));

    return function fetcher() {
      return _ref2.apply(this, arguments);
    };
  }();

  var _useSWR = useSWR("useQueryMetadata-".concat(id), fetcher),
      SWRData = _useSWR.data,
      isValidating = _useSWR.isValidating;

  useEffect(function () {
    var SWRValue = isSuccessResponse(SWRData) ? SWRData.value : {};
    var draftConfig = SWRValue.vis_config,
        draftModel = SWRValue.model,
        draftView = SWRValue.view;

    if (id && (!isEmpty(draftConfig) && !isEqual(metadata.vis_config, draftConfig) || draftModel && draftModel !== metadata.model || draftView && draftView !== metadata.view)) {
      var draftQuery = _objectSpread(_objectSpread({}, metadata), SWRValue);

      setById(id, {
        metadata: draftQuery
      });
    }
  }, [id, SWRData, metadata, setById]);
  return _objectSpread({
    isOK: !!metadata,
    isPending: isValidating,
    metadata: metadata
  }, getErrorResponse(SWRData));
};
